File: aic_data_collection/test/test_trial_config_provider.py
import pytest
import tempfile
import yaml

from aic_data_collection.trial_config_provider import StaticTrialProvider


@pytest.fixture
def sample_config_path():
    """Create a minimal sample config YAML for testing."""
    config = {
        "trials": {
            "trial_1": {
                "scene": {
                    "task_board": {"pose": {"x": 0.15, "y": -0.2, "z": 1.14, "roll": 0.0, "pitch": 0.0, "yaw": 3.14}},
                    "cables": {
                        "cable_0": {
                            "pose": {"gripper_offset": {"x": 0.0, "y": 0.015, "z": 0.042}, "roll": 0.44, "pitch": -0.48, "yaw": 1.33},
                            "attach_cable_to_gripper": True,
                            "cable_type": "sfp_sc_cable",
                        }
                    },
                },
                "tasks": {
                    "task_1": {
                        "cable_type": "sfp_sc", "cable_name": "cable_0",
                        "plug_type": "sfp", "plug_name": "sfp_tip",
                        "port_type": "sfp", "port_name": "sfp_port_0",
                        "target_module_name": "nic_card_mount_0", "time_limit": 180,
                    }
                },
            },
            "trial_2": {
                "scene": {
                    "task_board": {"pose": {"x": 0.15, "y": -0.2, "z": 1.14, "roll": 0.0, "pitch": 0.0, "yaw": 3.14}},
                    "cables": {
                        "cable_0": {
                            "pose": {"gripper_offset": {"x": 0.0, "y": 0.015, "z": 0.045}, "roll": 0.44, "pitch": -0.48, "yaw": 1.33},
                            "attach_cable_to_gripper": True,
                            "cable_type": "sfp_sc_cable",
                        }
                    },
                },
                "tasks": {
                    "task_1": {
                        "cable_type": "sfp_sc", "cable_name": "cable_0",
                        "plug_type": "sfp", "plug_name": "sfp_tip",
                        "port_type": "sfp", "port_name": "sfp_port_0",
                        "target_module_name": "nic_card_mount_1", "time_limit": 180,
                    }
                },
            },
        },
        "robot": {
            "home_joint_positions": {
                "shoulder_pan_joint": -0.1597,
                "shoulder_lift_joint": -1.3542,
                "elbow_joint": -1.6648,
                "wrist_1_joint": -1.6933,
                "wrist_2_joint": 1.5710,
                "wrist_3_joint": 1.4110,
            }
        },
    }
    with tempfile.NamedTemporaryFile(mode="w", suffix=".yaml", delete=False) as f:
        yaml.dump(config, f)
        return f.name


class TestStaticTrialProvider:
    def test_loads_all_trials(self, sample_config_path):
        provider = StaticTrialProvider(sample_config_path)
        assert provider.trial_count == 2

    def test_round_robin_default(self, sample_config_path):
        provider = StaticTrialProvider(sample_config_path)
        first = provider.get_next_trial()
        assert first["trial_id"] == "trial_1"
        second = provider.get_next_trial()
        assert second["trial_id"] == "trial_2"
        third = provider.get_next_trial()
        assert third["trial_id"] == "trial_1"  # wraps around

    def test_selected_trials(self, sample_config_path):
        provider = StaticTrialProvider(sample_config_path, trials=["trial_2"])
        first = provider.get_next_trial()
        assert first["trial_id"] == "trial_2"
        second = provider.get_next_trial()
        assert second["trial_id"] == "trial_2"  # only trial_2

    def test_trial_has_scene_and_tasks(self, sample_config_path):
        provider = StaticTrialProvider(sample_config_path)
        trial = provider.get_next_trial()
        assert "scene" in trial
        assert "tasks" in trial
        assert "cable_0" in trial["scene"]["cables"]

    def test_home_joint_positions(self, sample_config_path):
        provider = StaticTrialProvider(sample_config_path)
        home = provider.home_joint_positions
        assert home["shoulder_pan_joint"] == pytest.approx(-0.1597)

    def test_invalid_trial_name_raises(self, sample_config_path):
        with pytest.raises(ValueError, match="not found"):
            StaticTrialProvider(sample_config_path, trials=["trial_99"])

File: aic_engine/test/test_trial_config_provider.cpp
#include <gtest/gtest.h>
#include <fstream>
#include <cstdio>
#include <unistd.h>
#include "trial_config_provider.hpp"

static std::string write_sample_config() {
  const char* yaml = R"(
trials:
  trial_1:
    scene:
      task_board:
        pose: {x: 0.15, y: -0.2, z: 1.14, roll: 0.0, pitch: 0.0, yaw: 3.14}
      cables:
        cable_0:
          pose:
            gripper_offset: {x: 0.0, y: 0.015, z: 0.042}
            roll: 0.44
            pitch: -0.48
            yaw: 1.33
          attach_cable_to_gripper: true
          cable_type: sfp_sc_cable
    tasks:
      task_1:
        cable_type: sfp_sc
        cable_name: cable_0
        plug_type: sfp
        plug_name: sfp_tip
        port_type: sfp
        port_name: sfp_port_0
        target_module_name: nic_card_mount_0
        time_limit: 180
  trial_2:
    scene:
      task_board:
        pose: {x: 0.15, y: -0.2, z: 1.14, roll: 0.0, pitch: 0.0, yaw: 3.14}
      cables:
        cable_0:
          pose:
            gripper_offset: {x: 0.0, y: 0.015, z: 0.045}
            roll: 0.44
            pitch: -0.48
            yaw: 1.33
          attach_cable_to_gripper: true
          cable_type: sfp_sc_cable
    tasks:
      task_1:
        cable_type: sfp_sc
        cable_name: cable_0
        plug_type: sfp
        plug_name: sfp_tip
        port_type: sfp
        port_name: sfp_port_0
        target_module_name: nic_card_mount_1
        time_limit: 180
robot:
  home_joint_positions:
    shoulder_pan_joint: -0.1597
    shoulder_lift_joint: -1.3542
    elbow_joint: -1.6648
    wrist_1_joint: -1.6933
    wrist_2_joint: 1.571
    wrist_3_joint: 1.411
)";
  char path[] = "/tmp/aic_test_config_XXXXXX.yaml";
  int fd = mkstemps(path, 5);
  write(fd, yaml, strlen(yaml));
  close(fd);
  return std::string(path);
}

class StaticTrialProviderTest : public ::testing::Test {
protected:
  void SetUp() override { config_path_ = write_sample_config(); }
  void TearDown() override { std::remove(config_path_.c_str()); }
  std::string config_path_;
};

TEST_F(StaticTrialProviderTest, LoadsAllTrials) {
  aic::StaticTrialProvider provider(config_path_);
  EXPECT_EQ(provider.trial_count(), 2u);
}

TEST_F(StaticTrialProviderTest, RoundRobinDefault) {
  aic::StaticTrialProvider provider(config_path_);
  auto t1 = provider.get_next_trial();
  EXPECT_EQ(t1["trial_id"].as<std::string>(), "trial_1");
  auto t2 = provider.get_next_trial();
  EXPECT_EQ(t2["trial_id"].as<std::string>(), "trial_2");
  auto t3 = provider.get_next_trial();
  EXPECT_EQ(t3["trial_id"].as<std::string>(), "trial_1");
}

TEST_F(StaticTrialProviderTest, SelectedTrials) {
  aic::StaticTrialProvider provider(config_path_, {"trial_2"});
  auto t1 = provider.get_next_trial();
  EXPECT_EQ(t1["trial_id"].as<std::string>(), "trial_2");
  auto t2 = provider.get_next_trial();
  EXPECT_EQ(t2["trial_id"].as<std::string>(), "trial_2");
}

TEST_F(StaticTrialProviderTest, TrialHasSceneAndTasks) {
  aic::StaticTrialProvider provider(config_path_);
  auto trial = provider.get_next_trial();
  EXPECT_TRUE(trial["scene"].IsDefined());
  EXPECT_TRUE(trial["tasks"].IsDefined());
  EXPECT_TRUE(trial["scene"]["cables"]["cable_0"].IsDefined());
}

TEST_F(StaticTrialProviderTest, HomeJointPositions) {
  aic::StaticTrialProvider provider(config_path_);
  auto home = provider.home_joint_positions();
  EXPECT_NEAR(home["shoulder_pan_joint"], -0.1597, 1e-4);
}

TEST_F(StaticTrialProviderTest, InvalidTrialNameThrows) {
  EXPECT_THROW(
    aic::StaticTrialProvider(config_path_, {"trial_99"}),
    std::runtime_error
  );
}
